Remove PCC relocation from ISAv9

arichardson opened the pull request.

arichardson committed:
[CHERI-RISC-V] Legalize xTCC in CSpecialRW

If the new mode bits are reserved, we should retain the old mode
to match sail.

arichardson committed:
[CHERI-RISC-V][ISAv9] Use address instead of offset for SCRs

When ISAv9 semantics are enabled, reading/writing an SCR should affect
the address and not the offset of a capability.

arichardson committed:
[CHERI-RISC-V] Disable PCC relocation for ISAv9 semantics

PCC relocation has been removed from CHERI-RISC-V, so use a similar
approach to the change that removed DDC relocation from TCG.

arichardson committed:
Use the pcc_reloc(ctx) macro in more places

This allows removing two explicit `cctlr_set(ctx, CCTLR_PCCBO)` checks
from Morello since it's now handled by the macro.

bsdjhb commented on target/riscv/op_helper_cheri.c:
```diff
@@ -213,12 +213,34 @@ void HELPER(cspecialrw)(CPUArchState *env, uint32_t cd, uint32_t cs,
+        case CheriSCR_MTCC: {
+            target_ulong new_tvec = SCR_TO_PROGRAM_COUNTER(env, &new_val);
+            target_ulong new_mode = new_tvec & 3;
+            /* The low two bits encode the mode, but only 0 and 1 are valid. */
```
What mode does this refer to?  It doesn't seem to be present in the old code?

jrtc27 replied:
RISC-V's xtvec; 0 is direct, 1 is vectored, 2 is reserved but being claimed by extremely dodgy CLIC thing ("I can't believe it's not Arm-M"), 3 is reserved

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -1447,16 +1447,22 @@ void riscv_cpu_do_interrupt(CPUState *cs)
+         */
+        target_ulong current_offset = cap_get_offset(scr);
+        target_ulong diff = new_value - current_offset;
+        target_ulong new_cursor = cap_get_cursor(scr) + diff;
```
```suggestion
        new_cursor = cap_get_cursor(scr) + diff;
```

arichardson replied:
I just discovered this after spending 20 minutes debugging an untagged STCC... Should have just waited for your review :) Or turned on -Wshadow...

arichardson replied:
CI should fail with a more sensible error message now.

jrtc27 commented on target/riscv/translate.c:
```diff
@@ -486,7 +486,7 @@ static void gen_jal(DisasContext *ctx, int rd, target_ulong imm)
         }
     }
     // For CHERI the result is an offset relative to PCC.base
```
This comment is stale

arichardson committed:
[CHERI] Exit QEMU when we enter an inifinite CHERI trap loop

If the newly installed exception handler PCC is untagged and the same as
the previous PCC there is no way we can make forward progress, so exit
QEMU instead. This happened due to a variable shadowing bug that I
introduced as part of the removal of PCC relocation and was only
observable due to the timeouts in CI.
